src/puya/lib_embedded/puyapy_lib_arc4.py

```diff
@@ -0,0 +1,252 @@
+def dynamic_array_concat_variable_size(
+    source: Bytes, new_items_bytes: Bytes, new_items_count: UInt64
+) -> Bytes:
+    """
+    Concat new items to an arc4 dynamic array of variably sized items
+
+    source: The bytes of the source array
+    new_items_bytes: The bytes for all new items, concatenated
+    new_items_counts: The count of new items being added
+
+    returns: The updated bytes for the source array
+    """
+    array_length = extract_uint16(source, 0)
+    new_length = array_length + new_items_count
+    header_end = array_length * 2 + 2
+
+    return substring(itob(new_length), 6, 8) + recalculate_array_offsets_static(
+        array_data=(
+            substring(source, 2, header_end)
+            + bzero(new_items_count * 2)
+            + substring(source, header_end, source.length)
+            + new_items_bytes
+        ),
+        length=new_length,
+        start_at_index=UInt64(0),
+    )
+
+
+@subroutine
+def dynamic_array_concat_fixed_size(
+    source: Bytes, new_items_bytes: Bytes, new_items_count: UInt64
+) -> Bytes:
+    """
+    Concat new items to an arc4 dynamic array of fixed sized items
+
+    source: The bytes of the source array
+    new_items_bytes: The bytes for all new items, concatenated
+    new_items_counts: The count of new items being added
+
+    returns: The updated bytes for the source array
+    """
+    array_length = extract_uint16(source, 0)
+    source = replace(source, 0, substring(itob(array_length + new_items_count), 6, 8))
+    source += new_items_bytes
+    return source
+
+
+@subroutine
+def recalculate_array_offsets_dynamic(array_data: Bytes, start_at_index: UInt64) -> Bytes:
+    """
+    Recalculates the offset values of an arc4 dynamic array.
+
+    array_data: The dynamic array data
+    start_at_index: Optionally start at a non-zero index for performance optimisation. The offset
+                    at this index is assumed to be correct if start_at_index is not 0
+
+    returns: The updated bytes for the source array
+    """
+    return substring(array_data, 0, 2) + recalculate_array_offsets_static(
+        extract(array_data, 2, 0),
+        extract_uint16(array_data, 0),
+        start_at_index,
+    )
+
+
+@subroutine
+def recalculate_array_offsets_static(
+    array_data: Bytes, length: UInt64, start_at_index: UInt64
+) -> Bytes:
+    """
+    Recalculates the offset values of an arc4 static array.
+
+    array_data: The static array data
+    length: The length of the static array
+    start_at_index: Optionally start at a non-zero index for performance optimisation. The offset
+                    at this index is assumed to be correct if start_at_index is not 0
+
+    returns: The updated bytes for the source array
+    """
+    header_cursor = start_at_index * 2
+    if start_at_index == 0:
+        tail_cursor = length * 2
+    else:
+        tail_cursor = extract_uint16(array_data, header_cursor)
+
+    for i in urange(start_at_index, length):  # noqa: B007
```

Comment:
Can use _i here to avoid the ruff warning

Reply:
Ah nice - at one point we blocked variables starting with underscore but I guess that got removed.

Reply:
We still block `_` as it is treated as an Any type, but any other variable prefixed with _ doesn't cause an unused variable warning